draft/XXXX-configuration_handling/README.md

```diff
@@ -0,0 +1,202 @@
+on, to automatically marching linearly through a range of floating point values,
+to swapping out entire algorithmic implementations to compare different
+approaches. These types of variations will be expressible through specific
+meta-configuration syntax within the configuration file schema. Some examples
+follow.
+
+#### Combinatorial variation
+
+The most straightforward idea is to provide lists of values for each varying
+paramater to assume, and then to produce a stream of configurations visiting
+every ordered tuple of such values in their collective product space. For
+instance, the following hypothetical snippet of a configuration:
+
+```yaml
+a: 0
+b:
+  values: [1, 2, 3]
+c:
+  values: [4, 5]
+```
+
+would result in the following sequence of six ordered triples for `(a, b, c)`:
+
+```
+(0, 1, 4)
+(0, 1, 5)
+(0, 2, 4)
+(0, 2, 5)
+(0, 3, 4)
+(0, 3, 5)
+```
+
+Note that `a = 0` in all six tuples, while `b` and `c` vary combinatorially as
+you would expect.
+
+#### Lockstep variation
+
+Combinatorial variation provides "multiplicative" effects, but we may want a
+certain kind of "additive" effect as well. The idea here is to vary a set of
+parameters as a group. For example:
+
+```yaml
+a: 0
+b:
+  lockstep: foo
+  value: [1, 2, 3]
+c:
+  lockstep: foo
+  value: [4, 5, 6]
+```
+
+yielding the following `(a, b, c)` ordered triples:
+
+```
+(0, 1, 4)
+(0, 2, 5)
+(0, 3, 6)
+```
+
+Note that this requires specifying which "lockstep group" each parameter is a
+part of (identified by a string, in this case `foo`). Note also that `a = 0`
+once more, as it is specified as a plain value.
+
+#### Ranges
+
+We may want to support different types of range operators. The `value` field in
+the foregoing examples takes a YAML list as its argument, explicitly listing the
+values to take on. However, something like the following:
+
+```yaml
+a: 0
+b:
+  value:
+    sequence:
+      start: 1
+      stop: 3
+      step: 1
+c:
+  value:
+    sequence:
+      start: 4
+      stop: 6
+```
+
+would be the equivalent of the combinatorial example above (with the `step`
+parameter defaulting to `1`, as in the case of `c`).
+
+Other ranges can be implemented as needed.
+
+#### SMQTK algorithms
+
+Tinker Engine includes first-class support for SMQTK. A special
+meta-configuration option will enable SMQTK algorithms to be used as swappable
+values during configuration processing. For example:
+
+```yaml
+foo:
+  value:
+    - smqtk: MyAlgorithm.FooBar
+      config:
+        foo: 1
+        bar: 2
+    - smqtk: MyAlgorithm.FooBaz
+      config:
+        foo: 1
+        bar: 2
+```
+
+This will cause the `foo` value to take on a SMQTK algorithm instance as
+directed by the arguments. The runtime will first check to see that
+`MyAlgorithm.FooBar` and `MyAlgorithm.FooBaz` are existing, usable SMQTK
+algorithms as part of the validation phase.
+
+#### Schema and validation
+
+The meta-configuration schema will be formalized in such a way that the format
```

Comment:
Yep, I imagine we would create a JSON schema for this, and then generate documentation directly from the schema. I'm not an expert in JSON schema so we might need a bit of help with that.

Reply:
This is what I was using script config for since it wrapped it in python and you just needed a python dictionary to define them.  I'd prefer to keep the users in python rather than switching to JSON validation coding (to keep things simpler for the user).

Reply:
I'm not totally sure what you're suggesting--specifically, what is "JSON validation coding"? If you're thinking that I am going to put the burden of validating config files on the user, that's not my intent. All I'm suggesting here is that we (eventually) express the correct structure of a Tinker configuration via a JSON schema, which in turn would yield validation routines and possible documentation as a side effect.

Reply:
I was thinking like these complex syntax validation for JSON can be tiresome.  Using something like this: https://json-schema.org/draft/2019-09/json-schema-validation.html is pretty complex to know when we can just use python.

Reply:
We can discuss this when we meet but I'm still not sure what the objection is. I'm suggesting we (eventually) define the semantics of our configuration file with a JSON schema, which can then be used to validate the user's configuration file so that there's a lower chance of a misunderstanding, or strange behavior, or incorrect behavior at runtime. This is pretty standard stuff in a system like this.

If you're saying that putting in this validation infrastructure is tiresome, I don't know what to tell you: we get paid to do this tiresome stuff :slightly_smiling_face:. If you mean that you don't want the users to have to engage in this type of validation coding, then I agree 100%, and that's not something I'm suggesting.

AFAICT, tools like scriptconfig would be orthogonal to the notion of providing JSON validation.

In any case, I'm not saying we have to do this right now. It's just something that can make the system we're designing more robust and less error prone in the long run.

Reply:
See 72dd3dec4844361aa26afaf8147bc14b90934b29 for a bit more detail on this.

Reply:
Let move this to another RFC after changing the name of this rfc